File: platform/scripts/audit/t0_retrodiction/lib/dates.ts
/**
 * dates.ts — date-math primitives for the T-0 retrodiction-gate harness
 * (BRIEF_D3.md §F1 Lane T-0). No network, no astrology — pure calendar math
 * so it is trivially unit-testable offline (CONDUCTOR_PROTOCOL §3.1(b)).
 */

const MS_PER_DAY = 86_400_000

/** Parses a 'YYYY-MM-DD' string as a UTC midnight Date (avoids TZ drift). */
export function parseDate(iso: string): Date {
  const [y, m, d] = iso.slice(0, 10).split('-').map(Number)
  return new Date(Date.UTC(y, m - 1, d))
}

export function toIsoDate(d: Date): string {
  return d.toISOString().slice(0, 10)
}

export function addDays(d: Date, days: number): Date {
  return new Date(d.getTime() + days * MS_PER_DAY)
}

/** Signed day difference b - a. */
export function daysBetween(a: Date, b: Date): number {
  return Math.round((b.getTime() - a.getTime()) / MS_PER_DAY)
}

export function isWithinDays(center: Date, candidate: Date, windowDays: number): boolean {
  return Math.abs(daysBetween(center, candidate)) <= windowDays
}

export function clampDate(d: Date, min: Date, max: Date): Date {
  if (d.getTime() < min.getTime()) return min
  if (d.getTime() > max.getTime()) return max
  return d
}

File: platform/python-sidecar/scripts/kala_admission/dates.py
"""
dates.py -- date-math primitives for the D-3 ADMIT kernel-admission harness.

Direct Python port of
platform/scripts/audit/t0_retrodiction/lib/dates.ts (same function names,
same semantics -- UTC-midnight dates, signed day differences). No network,
no astrology -- pure calendar math.
"""
from __future__ import annotations

from datetime import date, timedelta


def parse_date(iso: str) -> date:
    """Parses a 'YYYY-MM-DD' string into a date object."""
    y, m, d = (int(x) for x in iso[:10].split("-"))
    return date(y, m, d)


def to_iso_date(d: date) -> str:
    return d.isoformat()


def add_days(d: date, days: int) -> date:
    return d + timedelta(days=days)


def days_between(a: date, b: date) -> int:
    """Signed day difference b - a."""
    return (b - a).days


def is_within_days(center: date, candidate: date, window_days: int) -> bool:
    return abs(days_between(center, candidate)) <= window_days


def clamp_date(d: date, dmin: date, dmax: date) -> date:
    if d < dmin:
        return dmin
    if d > dmax:
        return dmax
    return d
